Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm

```diff
@@ -249,7 +249,39 @@ static bool experimentalFeatureEnabled(const String& key, bool defaultValue = fa
         websiteDataStoreDirectory.get() = [webkitDirectory URLByAppendingPathComponent:@"WebsiteDataStore" isDirectory:YES];
     });
 
-    return [websiteDataStoreDirectory.get() URLByAppendingPathComponent:identifier isDirectory:YES].absoluteURL.path;
+    return websiteDataStoreDirectory.get().get().absoluteURL.path;
+}
+
+void WebsiteDataStore::fetchAllDataStoreIdentifiers(CompletionHandler<void(Vector<UUID>&&)>&& completionHandler)
+{
+    Vector<UUID> identifiers;
+    for (auto identifierString : FileSystem::listDirectory(defaultWebsiteDataStoreRootDirectory())) {
+        if (auto identifier = UUID::parse(identifierString))
+            identifiers.append(*identifier);
+    }
+
+    completionHandler(WTFMove(identifiers));
+}
+
+void WebsiteDataStore::deleteDataStoreWithIdentifier(const UUID& identifier, CompletionHandler<void(const String&)>&& completionHandler)
+{
+    bool isIdentifierInUse = false;
+    forEachWebsiteDataStore([&](auto& dataStore) {
+        if (auto storeIdentifier = dataStore.identifier(); storeIdentifier && *storeIdentifier == identifier)
+            isIdentifierInUse = true;
+    });
+    if (isIdentifierInUse)
+        return completionHandler("WebsiteDataStore with this identifier is in use"_s);
+
+    if (!FileSystem::deleteNonEmptyDirectory(defaultWebsiteDataStoreDirectory(identifier)))
+        return completionHandler("WebsiteDataStore with this identifier does not exist"_s);
```

Comment:
It could exist but the delete operation could fail.

Reply:
Will change the error message to reflect that.